docs: fix prose for extended expression (#434)
substrait-io/substrait: site/docs/expressions/extended_expression.md

# Extended expression

> Extended expression is a top-level message, like a plan, but for expressions. They can hold expressions to be used in filter, projection, and aggregation operations that aren't part of a full plan. Because they are a top-level message, they include more context than expressions that exist _within_ a plan. Some of this is additional computational context, such as the input data schema and the output schema. And some are message metadata, such as the Substrait version and the URIs of referenced extensions.

Reviewer: ```suggestion
Extended expression is a top-level message, like a plan, but for expressions. They can hold expressions to be used in filter, projection, and aggregation operations that aren't part of a full plan. Because it is a top-level message, it includes more context than expressions that exist _within_ a plan. Some of this is additional computational context, such as the input data schema and the output schema.  In addition there is message metadata, such as the Substrait version and the URIs of referenced extensions.
```

> Extended expressions are a top-level message, like a plan, but for expressions. They can be used to pass around expressions for use in filter, projection, and aggregation operations that aren't part of a full plan. Because they are a top-level message, they include more context than expressions that exist _within_ a plan. Some of this is additional computational context, such as the input data schema and the output schema. And some are message metadata, such as the Substrait version and the URIs of referenced extensions.

Reviewer 2: @wjones127 just a nit

```suggestion
Extended expressions are a top-level message, like a plan, but for expressions. They can pass around expressions to be used in filter, projection, and aggregation operations that aren't part of a full plan. Because they are a top-level message, they include more context than expressions that exist _within_ a plan. Some of this is additional computational context, such as the input data schema and the output schema. And some are message metadata, such as the Substrait version and the URIs of referenced extensions.
```

For details, see the definition in [extended_expression.proto](https://github.com/substrait-io/substrait/blob/main/proto/substrait/extended_expression.proto).

## Referred expression

> Extended expressions can contain one or multiple expressions. A single expression is used for filters, while multiple expressions are used for projections and aggregations. Each of these expressions is represented by an `ExpressionReference` message, which can either contain an [Expression](https://github.com/substrait-io/substrait/blob/7f272f13f22cd5f5842baea42bcf7961e6251881/proto/substrait/algebra.proto) or [AggregateFunction](https://github.com/substrait-io/substrait/blob/7f272f13f22cd5f5842baea42bcf7961e6251881/proto/substrait/algebra.proto#L1170) message. In the future, other types might be allowed to handle different use cases.

Reviewer: > A single expression is used for filters, while multiple expressions are used for projections and aggregations.

This seems overly prescriptive.  For example, I feel like this implies you would only use this for one of those three things.  However, someone might want to use this for expression evaluation that is completely unrelated to execution plans or engines.

Maybe something along the lines of "if multiple expressions are specified then there is no implicit relationship between the different top-level expressions.  Each expression should be considered independent and operates on the same top-level schema."

Author: Perhaps I can offer that as example usage, but not present as only valid use?

Author: But I will also add that note about the expressions not depending on each other.

> Extended expressions can contain one or multiple expressions. A single expression is used for filters, while multiple expressions are used for projections and aggregations. Each of these expressions are represented by an `ExpressionReference` messsage, which can either contain an [Expression](https://github.com/substrait-io/substrait/blob/7f272f13f22cd5f5842baea42bcf7961e6251881/proto/substrait/algebra.proto) or [AggregateFunction](https://github.com/substrait-io/substrait/blob/7f272f13f22cd5f5842baea42bcf7961e6251881/proto/substrait/algebra.proto#L1170) message. In the future, other types might be allowed to handle other use cases.

Reviewer 2: nit

```suggestion
Extended expressions can contain one or multiple expressions. A single expression is used for filters, while multiple expressions are used for projections and aggregations. Each of these expressions is represented by an `ExpressionReference` message, which can either contain an [Expression](https://github.com/substrait-io/substrait/blob/7f272f13f22cd5f5842baea42bcf7961e6251881/proto/substrait/algebra.proto) or [AggregateFunction](https://github.com/substrait-io/substrait/blob/7f272f13f22cd5f5842baea42bcf7961e6251881/proto/substrait/algebra.proto#L1170) message. In the future, other types might be allowed to handle different use cases.
```

## Input and output data schema

The schema of the input data is specified by the `base_schema` field.

The schema of the output data is determined by each of the `ExpressionReference` messages in the `referred_expr` field. The output type for the expression is determined by the expression or measure. The field name is determined by `output_names`. For primitive types `output_names` is just one string, but if the output type is a struct it will contain a name for each field, in depth-first order.

## Order

> If there are multiple expressions, the order of the inputs and outputs matter. Input data will match the schema order

Reviewer: What does "the order of the inputs matter" mean?

I agree that the output schema must align with the expression order.

Author: Changed to "the field order of the input and output data matters".

> in `base_schema` and output data must match the order of the `ExpressionReference` messages. However, there is no requirement that they expressions are executed in any particular order.

Reviewer 2: ```suggestion
in `base_schema` and output data must match the order of the `ExpressionReference` messages. However, there is no requirement that the expressions are executed in any particular order.
```

# Extended expression

> Extended expression is a top-level message, like a plan, but for expressions. They can hold expressions to be used in filter, projection, and aggregation operations that aren't part of a full plan. Because it is a top-level message, it includes more context than expressions that exist _within_ a plan. Some of this is additional computational context, such as the input data schema and the output schema.  In addition there is message metadata, such as the Substrait version and the URIs of referenced extensions.

Reviewer: ```suggestion
Extended expression is a top-level message, like a plan, but for expressions. It can hold expressions to be used in filter, projection, and aggregation operations that aren't part of a full plan. Because it is a top-level message, it includes more context than expressions that exist _within_ a plan. Some of this is additional computational context, such as the input data schema and the output schema.  In addition there is message metadata, such as the Substrait version and the URIs of referenced extensions.
```
